Autolev parser

NikhilPappu opened the pull request:
<!-- BEGIN RELEASE NOTES -->
* parsing
    * Added a submodule autolev which can be used to parse Autolev code to SymPy code.
* physics.mechanics
    * Added a center of mass function in functions.py which returns the position vector of the center of 
      mass of a system of bodies.
    * Added a corner case check in kane.py (Passes dummy symbols to q_ind and kd_eqs if not passed in 
       to prevent errors which shouldn't occur).
* physics.vector
    * Changed _w_diff_dcm in frame.py to get the correct results.
<!-- END RELEASE NOTES -->

asmeurer commented:
This should follow some of the same patterns as the LaTeX parser. Files that are autogenerated should contain a warning at the top that they are autogenerated and shouldn't be generated by hand.

asmeurer commented:
Also it should use the same version of antlr if possible, so that we don't have to install different versions for the tests.

NikhilPappu commented:
@asmeurer Seems like the antlr version I used (4.7.1) is the same as the one used in the LaTeX parser.
Were the LaTeX parser files generated with the target language Python 2? I am generating the antlr files with target language Python 3 and it is using type annotations (supposedly a Python 3.5 feature) everywhere which is failing on the Python 2 tests.

NikhilPappu commented:
@asmeurer I don't understand the Travis errors. Could you tell me why they are occurring?
There seems to be some error with test_setup.py
`$ bin/test_travis.sh
+[[ true == \t\r\u\e ]]
+python bin/test_setup.py
Traceback (most recent call last):
  File "bin/test_setup.py", line 26, in <module>
    assert setup.modules == module_list, set(setup.modules).symmetric_difference(set(module_list))
AssertionError: set([])`

Also the line `from antlr4 import *` in the antlr generated files is throwing an error regarding implicit imports. This line appears in the antlr files of the LaTeX parser as well. Why is that passing while this isn't?

certik commented:
I think it looks good. Overall I am against checking such large autogenerated files into git, as it increases the size of the git repository (forever). We can generate them for the release tarball.

Regarding tests, is this tested on Travis? It should be.

I see some test failures, would you mind fixing those please.

NikhilPappu commented:
@certik I'm not sure I quite understood what you meant. Do you mean the antlr generated files? How would this be different from the LaTeX parser which also has large autogenerated files? I think generating them only for a release tarball is a good idea though.

Did you mean the unit tests? I did add a file test_autolev.py which checks the parser generated outputs against the correct respective outputs for the example Autolev codes in the test_examples folder. I think this a good way to write tests as it is not easy to write simple one liner unit tests. For example the Kane() command needs the context of the whole system and can't be checked on its own.

I'll make sure the Travis build passes.

certik commented on sympy/parsing/autolev/Autolev.g4:
```diff
@@ -0,0 +1,111 @@
+
+vec:  ID ('>')+
+|      '0>'
+|      '1>>';
```
In here, please add spaces before "|", so that it is aligned with the ":", just like you have done for the `expr:` below. I think that seems to be the accepted formatting for the g4 files.

certik replied:
And the same for things above.

certik commented:
@NikhilPappu yes, I meant the antlr generated parser. Yes, it is the same for the latex parser. If you look at the files, they even have a binary part at the beginning (properly encoded as a Python string). I think in general it's best if such files are not checked into the git repository, but rather generated in the release tarball (or on Travis for tests). It does make the git repository not "self-contained" in a sense that you have to have antlr installed in order to generate those, but I think users should use the tarball (or Conda packages, or other packages) all of which would have it generated (since they are usually built from the tarball). /cc @asmeurer 

Regarding the tests in this PR, I think they way you did it is perfect. I agree with your point.

So as far as I am concerned, just do:

* fix the Travis tests
* change the formatting in the g4 based on my two comments above

NikhilPappu commented:
@certik I will change the formatting in the g4 file. Can you tell me why Travis is failing? I don't understand the reason. I ran bin/test parsing, bin/doctest parsing and bin/test quality in python 2 and 3 on my pc and they worked.

NikhilPappu commented:
@certik 
One error was due to this in the doctests and tests:
 >>> from sympy.parsing.autolev import parse_autolev
 >>> directory = "sympy/parsing/autolev/test_examples/"
 >>> parse_autolev(directory+"test7_in.txt", "print")`
Maybe this was a reason it passed on Windows but failed in Travis which uses Unix I think.
Will changing it to `parse_autolev(os.path.abspath(directory+"test7_in.txt"), "print")` work?
What would be the best way to read a file in a particular directory for the tests and doctests?

moorepants commented:
Make sure to use `os.path.join` to build directory strings.

moorepants commented on sympy/parsing/autolev/__init__.py:
```diff
@@ -0,0 +1,47 @@
+from sympy.external import import_module
+
+def parse_autolev(inp, outFile=None):
+    """
```
Please follow numpydoc standards for docstrings.

moorepants commented on sympy/parsing/autolev/__init__.py:
```diff
@@ -0,0 +1,47 @@
+
+from sympy.external import import_module
+
+def parse_autolev(inp, outFile=None):
```
use a PEP8 linter and follow its suggestions, for example: `outFile` -> `out_file`.

moorepants commented on sympy/parsing/autolev/test_examples/test10_in.txt:
```diff
@@ -0,0 +1,59 @@
+Newtonian N
```
Where do these examples come from? We need to assess the copyright and our use of them. Adding them to the sympy repo means they are licensed under the BSD license.

NikhilPappu replied:
All the examples added till now are from this Autolev Tutorial.
http://web.mae.ufl.edu/~fregly/PDFs/autolev_tutorial.pdf

NikhilPappu replied:
I would like to use some examples from Dynamics Online henceforth but as you said previously we might need to alter the code a little due to copyright issues.

moorepants replied:
Does the tutorial have any copyright information?

moorepants replied:
@certik Do you have any thoughts on how to handle this? I'm pretty sure both of these sources are not under any kind of open license.

NikhilPappu replied:
The first page says Copyright 1996-2005 by Paul Mitiguy and Keith Reckdahl. All Rights Reserved.

certik replied:
Yes, regarding the copyright, there are only two ways out. Either

* ask the authors to license the files we use under a BSD license

or

* rewrite the files from scratch (without referencing them), i.e., write them as good tests, without having any similarity to the original files. You can still use the original files on your local computer to ensure it actually works, but in sympy we have to have our own files.

moorepants commented on sympy/parsing/autolev/test_examples/test10_out.py:
```diff
@@ -0,0 +1,58 @@
+zero = sm.Matrix([i.collect(g)for i in zero]).reshape((zero).shape[0], (zero).shape[1])
+
+import scipy.integrate as sc
+import pydy.codegen.ode_function_generators as pd
```
This makes pydy a dependency of sympy, no? I don't think that we want any of the numerics based on pydy functions in sympy. SymPy is currently a strict dependency of PyDy, not the other way around. What you can do is generate numeric evaluation results using `evalf()`. You can compare the arbitrary precision values.

NikhilPappu replied:
I'm not sure what you mean by using evalf() here. Is it fine to use the other imports?
Is it fine to do it like you have done here : http://www.moorepants.info/blog/npendulum.html?
You don't seem to use a PyDy import here.

moorepants replied:
If you want to check numerical results in SymPy you need to use things like `xreplace`, `subs`, and `evalf`. This calls the underlying mpmath library and outputs arbitrary precision results. For long complicated expressions, evaluating them in both Autolev and SymPy numerically for random inputs is a good way to check the correctness.

You can use lambdify as in the npendulum blog post too, but this will give you floating point precision.

NikhilPappu replied:
I haven't written tests to compare the numerical outputs of SymPy and Autolev yet.
The current tests in test_autolev.py and parsing/test_examples just compare the parser generated output against the correct output so that if anything is changed one can see if anything is broken.

The code generated here is the code for Simulation of Eoms after the system has been specified.

moorepants commented:
This is looking great so far!.

Some big picture comments:

- This PR is way to big to review easily. For future PRs think about the smallest changes you can make and add them incrementally to build up functionality. For this one, don't add any more feature (reduce them if you can).
- It isn't clear where you do equivalency checks for autolev output and input.
- We decided to commit the autogenerated antlr files in the latex parser code, so that is what is done here. A separate PR could try to address not including these files and only having them generated in the sympy build process, as per Ondrej's suggestion. But for now I think we should follow suit.
- I don't recommend trying to get the bicycle or any other complicated examples working here. A PR per simple autolev example would be a better way to start. Build up the functionality incrementally.

moorepants commented:
We don't have an equivalent for `NICheck()`. But it seems like it wouldn't be hard to implement. The equation is creates is:
![selection_196](https://user-images.githubusercontent.com/276007/41359988-90f0393e-6ee0-11e8-87f5-80d88b846acb.png)
Pg 293 of Online Dynamics

it may be that our momentum methods on particle and body do this. not sure about dotting into the partial velocities. It might be a matter of something of this flavor:

```
sigma = 0
for body_or_particle in system_objects:
    sigma += body_or_particle.momentum().dot(body_or_particle.partial_velocity)
```

Maybe you can create a function that does what NICheck() does. We need a better more informative name though...

NikhilPappu commented:
@moorepants 

#### Some clarifications:
- As I have mentioned in one of the above replies, I am using test_examples/test1-10_in.txt files as the 
  input and checking the parser generated output against the files named test1-10_out.py which contain 
  the correct outputs. 
  You can have a look at parsing/tests/test_autolev.py.
 - I have checked that the numerical results for the examples used till now are same in both Autolev and 
   the generated SymPy code but I haven't written tests for it. The current tests just check the output 
   codes.
- I have parsed Autolev commands Kane(), Input, Output and Code Dynamics() using scipy and numpy 
  code for Simulation of eoms. You seemed to have confused them up with numerical tests I think.

#### What I plan to do next:
- I won't be adding any large amount of code for a while. I will work on cleaning things up a bit (fixing 
  the Travis errors, pep8 conventions, adding more comments, small code changes for the convenience of 
  use etc).
- I will focus on using more simple examples like you suggested and won't include Bicycle Autolev, 
  Autolev Tutorial examples 5.8 and 5.9 as they are more complicated.
- I will add Examples 5.2 and 5.7 followed by some simple examples from Kane's book (probably by 
  tweaking the names a little) after I clean up the code to complete this PR.
  Any code additions they will need should be minimal (as most parts of these examples can already be 
  parsed) so the PR won't change much.

- Should I leave the commented parts of Example 5.6 and 5.7 (have a look at them once in the Autolev 
  Tutorial) for later (the more difficult examples PR)?
  They use a command of the type Kane(F1, F2) which shuffles the equations to be solved for the 
  specified forces f1 and f2 instead of for the coordinates and speeds. They go on to plot these forces 
  with time.
  I am thinking I'll leave NiCheck for later as well.

certik commented:
Regarding the antlr generated files: the problem is that once this is merged, even if the files are removed later, the git repository will still contain them, and thus the size of it will be increased forever.

@asmeurer, what do you think we should do here?

NikhilPappu commented:
@moorepants I'll change the names of test1-10_out.py to expected_output1-10.py as the name seems misleading.

NikhilPappu commented:
@asmeurer @certik 
This is the kind of code I am using in test_autolev.py
```
directory = "sympy/parsing/tests/test_examples/"
parse_autolev(os.path.join(*(directory+inFileName).split('/')), os.path.join(*(directory+"output.py").split('/')))
correctFile = open(os.path.join(*(directory+outFileName).split('/')), 'r')
outputFile = open(os.path.join(*(directory+"output.py").split('/')), 'r')`
```
This is passing on my PC when I run bin/test parsing but it is causing Travis errors of this sort:
```
 File "/home/travis/virtualenv/python2.7.14/lib/python2.7/site-packages/sympy-1.1.2.dev0-py2.7.egg/sympy/utilities/runtests.py", line 1274, in _timeout
    function()
  File "/home/travis/virtualenv/python2.7.14/lib/python2.7/site-packages/sympy-1.1.2.dev0-py2.7.egg/sympy/parsing/tests/test_autolev.py", line 39, in test_example1
    _test_examples("test1_in.txt", "expected_output1.txt")
  File "/home/travis/virtualenv/python2.7.14/lib/python2.7/site-packages/sympy-1.1.2.dev0-py2.7.egg/sympy/parsing/tests/test_autolev.py", line 12, in _test_examples
    correctFile = open(os.path.join(*(directory+outFileName).split('/')), 'r')
IOError: [Errno 2] No such file or directory: 'sympy/parsing/tests/test_examples/expected_output1.txt'
```

How can I fix this? How can I **access specific files** in the SymPy directories?
Will using something like os.path.abspath() or os.getcwd() work?
I think fixing this would fix the Travis errors.
What would be the best way to do this type of thing?

moorepants commented:
You should use ``__file__`` to get relative locations to the currently executing file.

NikhilPappu commented:
@moorepants Can you show me an example? Also tests seem to be run using utilities/runtests.py in Travis while I use bin/test on my PC. Would that make any difference?

moorepants commented:
Does this help: https://stackoverflow.com/questions/5137497/find-current-directory-and-files-directory

NikhilPappu commented:
@moorepants That looks helpful. I'll try something like that and see if it works.

moorepants commented:
Also if you need to get ahold of a seperator, always use `os.path.sep`

moorepants commented:
You may want to do a tutorial on os.path, something like: https://pymotw.com/2/ospath/

NikhilPappu commented:
@moorepants Travis is throwing another error:
```
$ bin/test_travis.sh
+[[ true == \t\r\u\e ]]
+python bin/test_setup.py
Traceback (most recent call last):
  File "bin/test_setup.py", line 26, in <module>
    assert setup.modules == module_list, set(setup.modules).symmetric_difference(set(module_list))
AssertionError: set([])
```
How would I fix this?

NikhilPappu commented:
@certik @moorepants 
I've changed my tests. I don't need read and write to files anymore. However I am still getting errors even when importing stuff:
```
______________ sympy/parsing/tests/test_autolev.py:test_example1 _______________
  File "/home/travis/virtualenv/python2.7.14/lib/python2.7/site-packages/sympy-1.1.2.dev0-py2.7.egg/sympy/utilities/runtests.py", line 1274, in _timeout
    function()
  File "/home/travis/virtualenv/python2.7.14/lib/python2.7/site-packages/sympy-1.1.2.dev0-py2.7.egg/sympy/parsing/tests/test_autolev.py", line 15, in test_example1
    _test_examples(test1.autolev_code, test1.correct_sympy_code)
  File "/home/travis/virtualenv/python2.7.14/lib/python2.7/site-packages/sympy-1.1.2.dev0-py2.7.egg/sympy/parsing/tests/test_autolev.py", line 6, in _test_examples
    for i in range(min(len(generated_sympy_code), len(correct_sympy_code))):
TypeError: object of type 'NoneType' has no len()
```
These pass just fine on my PC.

This is another error I am getting:
```
$ bin/test_travis.sh
+[[ true == \t\r\u\e ]]
+python bin/test_setup.py
Traceback (most recent call last):
  File "bin/test_setup.py", line 26, in <module>
    assert setup.modules == module_list, set(setup.modules).symmetric_difference(set(module_list))
AssertionError: set([])

```
Can you go over my test files and all the Travis errors and tell me what I need to fix.
I am not sure what to do.

certik commented:
@NikhilPappu you have to make the test more robust, so that they provide good error messages. So from these lines:
```
  File "/home/travis/virtualenv/python2.7.14/lib/python2.7/site-packages/sympy-1.1.2.dev0-py2.7.egg/sympy/parsing/tests/test_autolev.py", line 6, in _test_examples
    for i in range(min(len(generated_sympy_code), len(correct_sympy_code))):
TypeError: object of type 'NoneType' has no len()
```
I can infer that either `generated_sympy_code` or `correct_sympy_code` is None, and then the `for i in range(...` line fails.

So you should modify your code to ensure that those two variables are never None. And whatever command sets them to None must return a readable error message to the user. That way you should be able to figure out easily why the code breaks on Travis.

NikhilPappu commented:
@certik 
This is the code I have in test_autolev.py:
```
from sympy.parsing.autolev.test_files import test1, test2, test3, test4, test5, test6, test7, test8, test9, test10
from sympy.parsing.autolev import parse_autolev
def _test_examples(autolev_code, correct_sympy_code):
    generated_sympy_code = parse_autolev("\n".join(autolev_code), "list")
    for i in range(min(len(generated_sympy_code), len(correct_sympy_code))):
        # print(generated_sympy_code[i] + "|  gen_code: " + str(i+1))
        # print(correct_sympy_code[i] + "|  corr_code: " + str(i+1))
        try:
            assert correct_sympy_code[i].rstrip() == generated_sympy_code[i].rstrip()
        except Exception:
            raise AssertionError("mismatch in line no: " + str(i+1))

def test_example1():
    _test_examples(test1.autolev_code, test1.correct_sympy_code)
```
I think its not able to import the files properly. Any reason you know of why this could be?

NikhilPappu commented:
@moorepants I am currently going through the chapters of Dynamics Online and making sure that the parser is able to parse most of the Autolev code given in the book.
I came across one thing that SymPy doesn't seem to be able to do.
When I parse code like this:
![autolev_diff2](https://user-images.githubusercontent.com/5346855/42055530-05779756-7b35-11e8-9b2f-98ec859d7b7d.PNG)
to something like this and run it:
![sympy_diff](https://user-images.githubusercontent.com/5346855/42055821-e9b0c230-7b35-11e8-8dff-20c142654bf5.PNG)

It gives me an error:
`ValueError: No Connecting Path found between a and b`
I want to fix this and I guess I have to implement something like this:
![autolev_diff](https://user-images.githubusercontent.com/5346855/42055671-726219fe-7b35-11e8-81ce-99b1013720cb.PNG)
Where in SymPy and how should I go about doing this?

NikhilPappu commented:
@moorepants @certik 
I have written a blog post about the status of the project and the plan for phase 3. Please go over it on Planet SymPy and give me feedback on it. I have added a commit now after parsing the first 3 chapters of Dynamics Online. The tests won't pass right now as I did not update them. I shall update them later.
Check out the file dynamics_online_codes.py. It contains all the Dynamics Online codes I have parsed till now along with comments and information about things like errors.

Abdullahjavednesar commented:
@NikhilPappu please resolve conflicts.

certik commented:
@NikhilPappu I would suggest to get this PR into a state that can be merged. As a start, let's get all tests to pass.

certik commented:
The PR looks good to me. As you know, I am against committing the ANTLR generated files, but @moorepants thinks it's ok and @asmeurer hasn't reacted yet.

@asmeurer are you ok with merging the big autogenerated files into sympy? I am personally against it, see my reasons above, but you are ok with it, we can merge it.

asmeurer commented:
Sorry, I'm busy at SciPy now, but I'll try to get a look at this soon.

certik commented:
@asmeurer thanks.

asmeurer commented:
It looks like the generated files aren't that big. They are smaller than the latex files. I am neural on including them. If you decide not to, you should rebase them out of this branch before merging so that they aren't in the git history either.

asmeurer commented:
I made some changes to the latex parser here to properly support case-insensitive file systems https://github.com/sympy/sympy/pull/14838/files#diff-83fcdf5d11d780bf2825d3500ae440ed. Do those need to be backported here? Can this share code with the LaTeX parser?

NikhilPappu commented:
@certik Sure. I just didn't update them till now because they tend to change a bit as I change my parser code. I think I'll just update them for now and make them pass. I could update them later as part of another PR again I guess Other than that ill probably have to clean up the code in _listener_autolev_antlr.py. _dynamics_online_codes.py is just something I want you to have a look at. I think it doesn't make much sense to merge that. I'll share that with you separately.

certik commented:
@asmeurer thanks for the feedback.

@NikhilPappu just do your best preparing this MR so that it can be merged. I am generally ok with what is here. If you have any doubts, let's discuss particular parts.

NikhilPappu commented:
@certik @asmeurer 
I have removed all the previous tests and have replaced them with a self made Autolev example using the mass spring damper model from the PyDy example repo. I shall make a few more examples like these using the double pendulum example, chaos pendulum example etc.

I think this is the best thing to do as for the tests to be sufficient I would have to include all the codes from Dynamics Online and the Autolev Tutorial which would be a lot in number (60+ right now).
So what I have decided to do is to put all of these test examples in another repo. Users will be able to directly copy these into the parsing/autolev/test_examples folder and use these tests.

I will decide whether to make the repo for the additional tests public or private based on the response from MotionGenesis.

certik commented:
I think you should write your own tests, that exercise all of your parser. I don't think that's hard, you just have to exercise each rule and you can write such tests yourself. Just keep it as short as possible, but do exercise all your code. Otherwise it will be very hard to make modifications to your code and ensure things still work.

The extra tests are good to have, in case your simpler tests didn't catch some corner case. But in this PR, let's add enough tests so that all your code is tested.

NikhilPappu commented:
I think that is a good idea Ondrej. I'll work on writing tests for all the
rules.

On Tue 17 Jul, 2018, 3:19 AM Ondřej Čertík, <notifications@github.com>
wrote:

> I think you should write your own tests, that exercise all of your parser.
> I don't think that's hard, you just have to exercise each rule and you can
> write such tests yourself. Just keep it as short as possible, but do
> exercise all your code. Otherwise it will be very hard to make
> modifications to your code and ensure things still work.
>
> The extra tests are good to have, in case your simpler tests didn't catch
> some corner case. But in this PR, let's add enough tests so that all your
> code is tested.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/sympy/sympy/pull/14758#issuecomment-405392647>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AFGWJxXCqG-sfZb73n5exq1qCEkofvrdks5uHQqBgaJpZM4UT76l>
> .
>

certik commented:
@NikhilPappu great. Ping me once you want me to review again.

asmeurer commented:
@NikhilPappu you should be writing better commit messages. "Small change" doesn't tell the reader anything about what is in the commit. See https://github.com/sympy/sympy/wiki/Development-workflow#writing-commit-messages for some tips. I recommend avoiding the `-m` flag and writing the message in a text editor, so that you aren't tempted to make it too short.

asmeurer commented:
Does the git history of this branch contain files that are not licensed properly to be included in SymPy? If so, they should be rebased out, so that even the git history does not have copyright violations. Let me or Ondrej know if you need help doing this.

Since you will be rebasing anyway, this is also a good opportunity to clean up the git history. At the very least, you should improve all the commit messages (select `edit` when doing an interactive rebase). Smaller commits can also be squashed into one another, though this is less important than just having good messages. 

To whoever merges this branch, please do not merge until we have verified that no commit contains copyrighted files.

asmeurer reviewed (dismissed):
Please do not merge this branch until it has been verified that no commits in the history contain copyrighted files (the test files from Autolev).

asmeurer commented:
You can wait for the MotionGenesis response before doing this, but the commit message cleanup should happen regardless.

NikhilPappu commented:
@asmeurer @certik Sorry about the commits. I'll change the messages right away and merge the very small ones. I'll also remove the Autolev Tutorial examples from the commits as I have decided not to include them here anyway. I am writing my own tests for all the rules which will be included here.
I shall make all the other test examples (Autolev Tutorial, Dynamics Online, PyDy repo examples) available in another repo. These are a lot in number and will increase the space of SymPy by a lot if I put them here so it is better to put them in another repo. Users will be able to copy these files to the additional_tests folder and use them. I shall make this repo private (planning to use BitBucket) for now. I will change it to public if MotionGenesis approves of it.

NikhilPappu commented:
@asmeurer I seemed to have cloned the repo once again at some point in between. I shouldn't have done that. Now in rebase I have all these other commits in between my own ones. What do I do about this?

certik commented:
Yes, go ahead and squash it into one commit.

NikhilPappu commented:
@certik It added files from other commits as well for some reason.
How do I change it so that it only adds the autolev_parser files?
Sorry I am kinda confused with git rebase as I haven't used it before.

NikhilPappu commented:
@certik Can I start with the latest sympy clone, make the changes and force push to this branch?

NikhilPappu committed:
Added a parser to parse Autolev code to SymPy code.

NikhilPappu commented:
@certik I have pushed the commit. This is what I'll be doing next:

1. I shall write the tests for all rules and I shall include them here(in parsing/autolev/test_examples).
    I guess the PR will be ready for merging after this. In the meanwhile you can go over the existing code 
    if you like.
2.  I shall then go about parsing additional examples and improving the parser.
     I shall put all the additional test examples in another PR.
     That repo will contain folders like:
     - PyDy repo examples (I will add files like mass_spring_damper, double pendulum, bicycle model 
       etc here)
     - Dynamics Online codes (I will add codes of this book separated into 5 chapter folders).
     - Autolev Tutorial examples (I will add examples 5.1- 5.7 of this tutorial here)
I will make this repo private on BitBucket if MotionGenesis doesn't approve.
I think this repo will be a nice place for full physics examples where the parser can be seen in action. Users will be able to copy these files to autolev/test_examples to use these tests.
I shall also write comments in all the test .al and .py files so it will be easier to identify issues and small details to cover in the documentation, issues and report.

certik commented:
@NikhilPappu I agree. In this PR, let's not modify the parser anymore, but let's write thorough tests for it, so that all your code is exercised. Then let's get it merged.

After that, let's create a new PR with improving the parser to be able to parse more things.

Please ping me once you want me to review.

NikhilPappu commented:
@certik Can you tell me how to make the Travis tests pass? This is the only error:
https://travis-ci.org/sympy/sympy/jobs/406016325

NikhilPappu commented:
@certik Can you please go over the Travis build and tell me what is wrong. I don't seem to understand.
This is the only one which is failing:
https://travis-ci.org/sympy/sympy/jobs/406147262

certik commented:
@asmeurer the sympy tests take way too long in my opinion, I think we test too many things needlessly. But that's a separate issue.

Are you ok with merging this PR even though there are some timeouts on Travis? We can probably set this test to "can fail". The reason is that this PR is way delayed, and it needs to be merged.

asmeurer commented:
No, the timeouts need to be fixed. Otherwise they will cause issues in every other pull request.

Also, the rebase issue (removing the copyrighting files) needs to be taken care of before this is merged.

certik commented:
@asmeurer that's what I was asking @NikhilPappu. So the timeouts are caused by this PR? That means that the tests take a long time. So that indeed must be fixed.

NikhilPappu commented:
@asmeurer @certik 
I don't think the timeouts are caused by this PR. The tests don't take long to run as they are just comparing some small files line by line.
Also, I have removed the copyrighted files from the PR. Previous commits shouldn't exist as I have force pushed to this branch.

asmeurer commented:
They likely are caused by this branch, because that build tests the parser stuff (that requires the antlr runtime). But like I said, it's already close to the limit, so probably this just pushes it over. 

We could also go into the specific tests that are run and try to trim them down, so that it only tests those tests that require external dependencies. Ideally, it each test should be marked somehow with metadata so that we could just set a flag in the runner to only run those tests that require numpy or whatever. Right now it just tests each submodule that uses the external dependency https://github.com/sympy/sympy/blob/d6ffcfd4e2bd28f5486e56035a38574864a42bd7/bin/test_travis.sh#L94. So it could be improved. 

How much more work is left for @NikhilPappu's project, aside from writing more tests? Is there time to work on this, or would it distract too much?

certik commented:
@asmeurer, given the fact that this is @NikhilPappu's first PR, I would like to merge this one soon, and then for Nikhil to send several more PRs. It will take several more days to finish this PR, and as such, unfortunately, there is no time to work on that.

asmeurer commented:
I think the easiest way to resolve this issue would be to go through the tests that are run in that build, and figure out which ones are taking the longest that don't actually use any external dependencies, and split them out so that they aren't run. I suspect this can work. For instance, we're running the entire matrices tests just for NumPy, even though there are probably only a couple of tests in the whole module that actually require NumPy.

certik commented:
Ok. That tests work should be done as a separate PR, and it should be merged into master before this PR.

certik commented:
@asmeurer how about we split this slow test into two, in this PR, so that we can merge it? Then we can try to speed things up in a separate PR.

asmeurer commented:
@NikhilPappu OK, I looked through the commits. So none of the test files that are in this PR are copyrighted (you wrote them all yourself)?

NikhilPappu commented:
@asmeurer @certik 
I plan to work on writing additional test examples in a private repo. This would serve the purpose of additional tests, showcasing the parser and will also make it easier to give specific insights when writing the documentation. I will include examples from Autolev Tutorial, Dynamics Online and the PyDy example repo here. I would need to go through a lot of codes for this and I need to improve the parser to get to parse these. This will take a while I'm afraid. After this I will write the documentation so I feel I cannot focus on the timeout issue at the moment.

asmeurer commented:
@certik  I agree a PR fixing the Travis build runtime should be separate from this one. 

I'm fine with splitting this, so long as Travis doesn't fail or timeout regularly in master. And so long as you can ensure that the new code does get tested eventually (i.e., if we merge just the code and then put the tests in later, we need to make sure those tests do eventually get in, or else the code will just end up in there without tests).

NikhilPappu commented:
@asmeurer I wrote all these test files myself. There are no copyrighted files here. The tests aren't complete though. I need to write some more. I will be done with that in the next 2 days. I think that should be it for this PR.

certik commented:
@NikhilPappu, @asmeurer as part of this PR, let's concentrate on getting all tests (that you write yourself Nikhil) in. The goal of the tests should be that they run fast and that they test as much of your code as possible. Then I can help with splitting the Travis test into two, so that we can get this merged. 

Your plan Nikhil is good. Let's concentrate on finishing up the tests and getting this merged. Only then let's work on other stuff.

NikhilPappu commented:
@certik 
Yes, the plan is to finish up the tests and the PR before moving onto other stuff.

asmeurer commented:
OK, thanks @NikhilPappu, that was my main concern with this PR. My other concern now is the Travis tests timing out, because that will cause problems for every other PR.

NikhilPappu commented:
@asmeurer Would you please be able to look at it? I am not too familiar with stuff related to Travis builds and it will be difficult for me to focus on this. I don't think the tests themselves could be improved though. They are pretty simple and finish almost instantaneously when run locally. Its probably the antlr dependency and its probably just pushing it over the limit.

asmeurer commented:
@NikhilPappu I just tested on my computer and the tests and doctests combined for autolev take less than 2 seconds to run. Can you confirm this? If this is the case, then I don't think it is your fault that the tests timed out (probably like I said, they are already close to the limit, and the run took a little longer than usual). 

Let's see if they time out again in the [current build](https://travis-ci.org/sympy/sympy/jobs/406426786). If not, then this is probably just an issue that already exists in master, and this PR doesn't make it worse. In that case, just open an new issue for it, and we can get it fixed.

Unless I am missing something and it really does take more than 2 seconds to run the new tests here on Travis.

NikhilPappu commented:
@asmeurer Yes it just takes 2 seconds to run. I think Travis passed through that build this time around.

certik commented:
@NikhilPappu great. That is the goal, for the tests to run quickly. So I think just concentrate on finishing the tests. Don't worry about Travis in this PR.

certik commented:
Let's discuss general Travis issues in #14952.

NikhilPappu commented:
@certik I have finished writing the tests. Travis seems to have a similar problem again. One build is failing due to timeout.

NikhilPappu commented:
@certik I have finished writing the tests. I think the PR is good enough to be merged now.

certik reviewed (changes requested):
The changes now look good, tests pass. But the commits don't make much sense, for example commits like:

Just checking if the Travis tests pass

Checking to see if the Travis tests pass 

Those are not good commits or commit messages.

Also, in the commit https://github.com/sympy/sympy/pull/14758/commits/a59e4075d4b30cbced2224ecd38ff7eef176907a you unfortunately first committed some binary blobs and then removed them.

For all these reasons, can you please squash all your 9 commits into just one, and write a nice commit message?

NikhilPappu commented:
@certik I'll squash all the commits into one. The reason for those strange commit messages is that the Travis tests were sometimes timing out and passing some other times without any significant changes.

asmeurer commented:
In the future, do not squash large pull requests like this into one commit. The release notes entry shows four distinct changes, so there should be distinct commits. At the very least, your commit message should be at least as descriptive as your release notes (see https://github.com/sympy/sympy/wiki/Development-workflow#writing-commit-messages).

asmeurer commented:
Also please take a look at the [bot status](https://github.com/sympy/sympy/pull/14758#issuecomment-407546302) whenever you edit the release notes or before merging to verify that they look correct. In this case, there was a bug in the way it handled the multiple bullet points (which I have fixed at https://github.com/sympy/sympy-bot/pull/15).

moorepants commented on sympy/parsing/autolev/__init__.py:
```diff
@@ -0,0 +1,94 @@
+
+    Parameters
+    ----------
+    inp: str
```
Why not "input" instead of "inp"?

moorepants commented on sympy/parsing/autolev/__init__.py:
```diff
@@ -0,0 +1,94 @@
+            1. Can be the name of an output file to which the SymPy code should be written to.
+            2. Can be the string "print". In this case the SymPy code is written to stdout.
+            3. Can be the string "list". In this it returns a list containing the SymPy code.
+                Each element in the list corresponds to one line of code.
```
What happens when `output=None`? Maybe the default should be to stdout.

moorepants commented on sympy/parsing/autolev/__init__.py:
```diff
@@ -0,0 +1,94 @@
+    >>> l.append("INPUT Q1=.1,Q2=.2,U1=0,U2=0)"  # doctest: +SKIP
+    >>> l.append("INPUT TFINAL=10, INTEGSTP=.01)"  # doctest: +SKIP
+    >>> l.append("CODE DYNAMICS() double_pendulum.c)"  # doctest: +SKIP
+    >>> parse_autolev("\\n".join(l), "print")  # doctest: +SKIP
```
Any reason this can't be written like:

```pycon
>>> """\
... My
... Autolev
... code\
... """
...
```
```

?

NikhilPappu replied:
It can be written in a better way. I shall change it.

moorepants commented on sympy/parsing/autolev/__init__.py:
```diff
@@ -0,0 +1,94 @@
+    kane = me.KanesMethod(frame_n, q_ind=[q1,q2], u_ind=[u1, u2], kd_eqs = kd_eqs)
+    fr, frstar = kane.kanes_equations([particle_p, particle_r], forceList)
+    zero = fr+frstar
+    from pydy.system import System
```
I think the PyDy code output should be optional. We are creating a circular dependency of sorts by including it. Either the autolev parser belongs in PyDy or an extension to the parser belongs in PyDy. I think we should stick to outputting only what SymPy can do with SymPy code.

Another option could be a flag: `parse_autolev(..., include_pydy=True)`.

NikhilPappu replied:
I think adding the flag is a good idea. I shall do that.

moorepants commented on sympy/parsing/autolev/__init__.py:
```diff
@@ -0,0 +1,94 @@
+from sympy.external import import_module
+
+
+def parse_autolev(inp, output=None):
```
What version(s) of autolev code does this parse?

NikhilPappu replied:
I have followed the Autolev Tutorial and the Autolev version you sent me (these are version 4.1) and Dynamics Online (which uses an older version) as guides. I didn't find too much of a difference when I used the older codes in the newer version but I think I went with the newer version in case of a conflict or deprecation warning. So I would say 4.1.

moorepants replied:
Ok, stating clearly that we are following the autolev 4.1 spec is good.

moorepants commented on sympy/parsing/autolev/_parse_autolev_antlr.py:
```diff
@@ -0,0 +1,38 @@
+                                 __import__kwargs={'fromlist': ['AutolevLexer']}).AutolevLexer
+    AutolevListener = import_module('sympy.parsing.autolev._antlr.autolevlistener',
+                                    __import__kwargs={'fromlist': ['AutolevListener']}).AutolevListener
+except Exception:
```
This should be a specific exception.

NikhilPappu replied:
I copied this style from the LaTeX parser.
I think ImportError should work though.

moorepants commented on sympy/physics/vector/frame.py:
```diff
@@ -260,7 +260,9 @@ def _w_diff_dcm(self, otherframe):
-        angvelmat = diffed * dcm2diff.T
+        # angvelmat = diffed * dcm2diff.T
+        # This one seems to produce the correct result when I checked using Autolev.
+        angvelmat = dcm2diff*diffed.T
```
This should be in a separate PR with a specific test for it. What is the test case that is broken?

NikhilPappu replied:
I shall add a test for it.
There was no test case for this to begin with so I didn't break anything.

moorepants replied:
This change should not have been merged. Just wanted to note that this broke working code as seen here: https://github.com/sympy/sympy/issues/16824. This PR seemed to have been merged without my concern addressed.

moorepants commented on sympy/physics/mechanics/functions.py:
```diff
@@ -381,6 +382,55 @@ def gravity(acceleration, *bodies):
     return gravity_force
 
+
+def center_of_mass(point, *bodies):
```
Are there any tests for this?

NikhilPappu replied:
I haven't added a test for this. I shall do so in the next PR.

moorepants replied:
This was also not addressed before merging.

moorepants commented:
@NikhilPappu I've had some time to look at this. Sorry it is after it was merged. This is really nice work and a great addition. I've left some comments that can be addressed in new PRs.

Also, I'm a bit confused on how the testing works. What I imagine as a good test for this is that we write some autolev code, run the autolev code to get its symbolic results, then we run the parser on the autolev code to generate the sympy code, and then run the sympy code to get its symbolic results.

Once we have the symbolic results from both programs we can compare them in two ways:

1. Check whether the symbolics are the same. This would involved converting autolev results to sympy expressions and then subtracting the results from the parsed code and then simplifying to see if we get zero.

Or

2. We substitute in random arbitrary precision or floating point numbers into the symbolic expressions from both systems and compare the numerical results.

It isn't clear to me if either of these are done. And thus I am not sure how to tell whether the autolev parser generates equivalent code. Can you comment on this?

certik commented:
@moorepants thanks for the review. Yes, @NikhilPappu please address the stuff in new PRs.

To answer your question about tests --- my understanding is that it tests the (current) results from the parser. So if you change some code in the parser, it will break the tests, thus ensuring that the parser works. This has the advantage that it runs quickly (no pydy necessary). If a bug is discovered, i.e., that the tests, as currently written, test an incorrect result, then we correct the test and fix the bug.

We can always write slower and more deeper tests to actually run pydy and autoleve side by side and compare the results. But the above seems like an excellent way to test the parser itself.

moorepants commented:
I see how these are regression tests. But it isn't clear to me how we know the parser emits correctly (symbolic and numerically speaking) code.

certik commented:
@moorepants, yes, these are regression tests. We do not know that the parser is emitting correct code. In order to know that, we need to finish some tests in the pydy I assume. I would be nice if @NikhilPappu could write such tests also, probably in pydy, before the GSoC is over. At least a few examples.

moorepants commented:
I'll discuss with him and I should be able to help with some.

certik commented:
We can revert the broken change.

On Tue, Oct 15, 2019, at 9:46 PM, Jason K. Moore wrote:
> *@moorepants* commented on this pull request.
> 
> In sympy/physics/vector/frame.py <https://github.com/sympy/sympy/pull/14758#discussion_r335276062>:
> 
> > @@ -260,7 +260,9 @@ def _w_diff_dcm(self, otherframe):
>  from sympy.physics.vector.functions import dynamicsymbols
>  dcm2diff = self.dcm(otherframe)
>  diffed = dcm2diff.diff(dynamicsymbols._t)
> - angvelmat = diffed * dcm2diff.T
> + # angvelmat = diffed * dcm2diff.T
> + # This one seems to produce the correct result when I checked using Autolev.
> + angvelmat = dcm2diff*diffed.T
> This change should not have been merged. Just wanted to note that this broke working code as seen here: #16824 <https://github.com/sympy/sympy/issues/16824>. This PR seemed to have been merged without my concern addressed.
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub <https://github.com/sympy/sympy/pull/14758?email_source=notifications&email_token=AAAFAWDHGR5ALWM7TUIJVB3QO2MEBA5CNFSM4FCPX2S2YY3PNVWWK3TUL52HS4DFWFIHK3DMKJSXC5LFON2FEZLWNFSXPKTDN5WW2ZLOORPWSZGOCICQN5I#discussion_r335276062>, or unsubscribe <https://github.com/notifications/unsubscribe-auth/AAAFAWCKMH2HYMSEJPHPMJ3QO2MEBANCNFSM4FCPX2SQ>.
>

moorepants commented:
It was fixed in PR #16828. The issue is that I'm teaching a course with the package now and the bug is present in the latest release of SymPy which is what my students are using.

moorepants commented on sympy/physics/mechanics/kane.py:
```diff
@@ -121,6 +121,9 @@ def __init__(self, frame, q_ind, u_ind, kd_eqs=None, q_dependent=None,
         """Please read the online documentation. """
+        if not q_ind:
+            q_ind = [dynamicsymbols('dummy_q')]
+            kd_eqs = [dynamicsymbols('dummy_kd')]
```
This change was also introduced with no tests or explanation.

certik commented:
> It was fixed in PR #16828. The issue is that I'm teaching a course with the package now and the bug is present in the latest release of SymPy which is what my students are using.

I merged it, I am really sorry that it broke things. I am glad it is fixed. What are the possible paths forward to get it working for your students? I know @asmeurer is trying to do a new release, but that might not be on time for your classes. We can upload a conda package into a special channel just for you, using the latest master. Let me know what you want to do.

moorepants commented:
I'm running  a new jupyterhub server for the class. I can make a new docker image with the patch to make it available to the students. Don't worry about it. I just wanted to document what happened here and review the other changes.

certik commented:
Ok, perfect. Thank you.

On Wed, Oct 16, 2019, at 6:34 AM, Jason K. Moore wrote:
> I'm running a new jupyterhub server for the class. I can make a new docker image with the patch to make it available to the students. Don't worry about it. I just wanted to document what happened here and review the other changes.
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub <https://github.com/sympy/sympy/pull/14758?email_source=notifications&email_token=AAAFAWDBTPHREAP3AWNDY4TQO4J43A5CNFSM4FCPX2S2YY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOEBMP3XQ#issuecomment-542703070>, or unsubscribe <https://github.com/notifications/unsubscribe-auth/AAAFAWABRWZRDYHSNPKQP23QO4J43ANCNFSM4FCPX2SQ>.
>
